onnxruntime/core/providers/cuda/math/unary_elementwise_ops_impl.cu

```diff
@@ -115,50 +119,139 @@ struct OP_Cast {
-  SPECIALIZED_CAST_IMPL2(T, double)   \
-  SPECIALIZED_CAST_IMPL2(T, int8_t)   \
-  SPECIALIZED_CAST_IMPL2(T, int16_t)  \
-  SPECIALIZED_CAST_IMPL2(T, int32_t)  \
-  SPECIALIZED_CAST_IMPL2(T, int64_t)  \
-  SPECIALIZED_CAST_IMPL2(T, uint8_t)  \
-  SPECIALIZED_CAST_IMPL2(T, uint16_t) \
-  SPECIALIZED_CAST_IMPL2(T, uint32_t) \
-  SPECIALIZED_CAST_IMPL2(T, uint64_t) \
-  SPECIALIZED_CAST_IMPL2(T, bool)     \
-  SPECIALIZED_CAST_IMPL2(T, BFloat16)
-
-SPECIALIZED_CAST_FROM(half)
-SPECIALIZED_CAST_FROM(float)
-SPECIALIZED_CAST_FROM(double)
-SPECIALIZED_CAST_FROM(int8_t)
-SPECIALIZED_CAST_FROM(int16_t)
-SPECIALIZED_CAST_FROM(int32_t)
-SPECIALIZED_CAST_FROM(int64_t)
-SPECIALIZED_CAST_FROM(uint8_t)
-SPECIALIZED_CAST_FROM(uint16_t)
-SPECIALIZED_CAST_FROM(uint32_t)
-SPECIALIZED_CAST_FROM(uint64_t)
-SPECIALIZED_CAST_FROM(bool)
-SPECIALIZED_CAST_FROM(BFloat16)
+struct OP_CastNoSat {
+  __device__ __inline__ OutT operator()(const InT& a) const;
+};
+
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11080
+
+#define OP_CAST(T, NVT) \
+  template <> \
+  struct OP_CastSat<half, T> { \
+    __device__ __inline__ T operator()(const half& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_halfraw_to_fp8(v, __NV_SATFINITE, NVT)), T::FromBits()); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastNoSat<half, T> { \
+    __device__ __inline__ T operator()(const half& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_halfraw_to_fp8(v, __NV_NOSAT, NVT)), T::FromBits()); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastSat<float, T> { \
+    __device__ __inline__ T operator()(const float& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_float_to_fp8(v, __NV_SATFINITE, NVT)), T::FromBits()); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastNoSat<float, T> { \
+    __device__ __inline__ T operator()(const float& v) const { \
+      return T(static_cast<unsigned char>(__nv_cvt_float_to_fp8(v, __NV_NOSAT, NVT)), T::FromBits()); \
+    } \
+  };
+
+#else
+
+#define OP_CAST(T, NVT) \
+  template <> \
+  struct OP_CastSat<half, T> { \
+    __device__ __inline__ T operator()(const half& v) const { \
+      return T(__half2float(v), true); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastNoSat<half, T> { \
+    __device__ __inline__ T operator()(const half& v) const { \
+      return T(__half2float(v), false); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastSat<float, T> { \
+    __device__ __inline__ T operator()(const float& v) const { \
+      return T(v, true); \
+    } \
+  }; \
+  template <> \
+  struct OP_CastNoSat<float, T> { \
+    __device__ __inline__ T operator()(const float& v) const { \
+      return T(v, false); \
+    } \
+  };
+
+#endif
+
+OP_CAST(Float8E4M3FN, __NV_E4M3)
+OP_CAST(Float8E5M2, __NV_E5M2)
+
+#define EXPLICIT_IMPL_CASTSAT(InT, OutT) \
+  void _Explicit_Impl_CastSat(cudaStream_t stream, const InT* input_data, OutT* output_data, size_t count, bool saturate) { \
```

Comment:
valid identifier name would be `Explicit_impl_CastSat`

Reply:
renamed